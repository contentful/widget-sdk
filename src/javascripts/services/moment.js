'use strict';

angular.module('cf.utils')
.factory('moment', ['raw/moment', function (moment) {
  moment.locale('en', {
    calendar: {
      lastDay : '[Yesterday], LT',
      sameDay : '[Today], LT',
      nextDay : '[Tomorrow], LT',
      lastWeek : 'ddd, LT',
      nextWeek : '[Next] ddd, LT',
      sameElse : 'll'
    }
  });

  return moment;
}]);
