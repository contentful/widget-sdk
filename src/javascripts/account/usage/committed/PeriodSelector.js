import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import { Select, Option } from '@contentful/forma-36-react-components';
import { cond, constant, stubTrue } from 'lodash';

import { periodPropType } from './propTypes.es6';

const formatDate = date => moment(date).format('DD MMM');

export default class PeriodSelector extends React.Component {
  static propTypes = {
    periods: PropTypes.arrayOf(periodPropType).isRequired,
    selectedPeriodIndex: PropTypes.number.isRequired,
    onChange: PropTypes.func.isRequired
  };

  render() {
    const { periods, onChange, selectedPeriodIndex } = this.props;
    return (
      <span className="usage__period-selector">
        <label>Usage period</label>
        <Select
          defaultValue={selectedPeriodIndex}
          onChange={onChange}
          width="auto"
          name="period-selector"
          id="period-selector">
          {periods.map(({ startDate, endDate }, index) => {
            const isCurrentPeriod = endDate === null;
            const start = moment(startDate);
            const end = isCurrentPeriod
              ? moment(start)
                  .add(1, 'month')
                  .subtract(1, 'day')
              : moment(endDate);

            const value = `${index}`;

            return (
              <Option key={index} value={value}>{`${formatDate(start)} â€“ ${formatDate(end)} ${cond([
                [constant(isCurrentPeriod), constant('(current)')],
                [constant(end.year() === moment().year()), constant(`(${moment().to(start)})`)],
                [constant(stubTrue), constant(end.year())]
              ])()}`}</Option>
            );
          })}
        </Select>
      </span>
    );
  }
}
