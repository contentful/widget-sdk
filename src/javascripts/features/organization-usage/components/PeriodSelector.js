import React from 'react';
import moment from 'moment';
import { Select, Option } from '@contentful/forma-36-react-components';
import { cond, constant, stubTrue } from 'lodash';
import { track } from 'analytics/Analytics';
import { useUsageState, useUsageDispatch } from '../hooks/usageContext';
import { pick } from 'lodash/fp';
import { loadPeriodData } from '../services/UsageService';

const formatDate = (date) => moment(date).format('DD MMM');

export const PeriodSelector = () => {
  const { periods, selectedPeriodIndex, orgId, isTeamOrEnterpriseCustomer } = useUsageState();
  const dispatch = useUsageDispatch();

  const handleChange = async (e) => {
    dispatch({ type: 'SET_LOADING', value: true });

    const newPeriodIndex = parseInt(e.target.value);

    // analytics
    const oldPeriod = periods[selectedPeriodIndex];
    const newPeriod = periods[newPeriodIndex];
    track('usage:period_selected', {
      oldPeriod: pick(['startDate', 'endDate'], oldPeriod),
      newPeriod: pick(['startDate', 'endDate'], newPeriod),
    });

    dispatch({ type: 'CHANGE_PERIOD', value: newPeriodIndex });
    dispatch({ type: 'SWITCH_MAIN_TAB', value: 'apiRequest' });
    const data = await loadPeriodData(orgId, newPeriod);
    dispatch({ type: 'SET_USAGE_DATA', value: data });
    dispatch({ type: 'SET_LOADING', value: false });
  };

  const processDate = (startDate, endDate) => {
    const isCurrentPeriod = endDate === null || moment().diff(moment(endDate), 'days') == 0; // the end day is today
    const start = moment(startDate);
    const end = isCurrentPeriod
      ? moment(start).add(1, 'month').subtract(1, 'day')
      : moment(endDate);
    return `${formatDate(start)} â€“ ${formatDate(end)} ${cond([
      [constant(isCurrentPeriod), constant('(current)')],
      [constant(end.year() === moment().year()), constant(`(${moment().to(start)})`)],
      [constant(stubTrue), constant(end.year())],
    ])()}`;
  };

  const renderSelect = () => {
    return (
      <Select
        defaultValue={selectedPeriodIndex}
        onChange={handleChange}
        width="auto"
        name="period-selector"
        testId="period-selector"
        id="period-selector">
        {periods.map(({ startDate, endDate }, index) => {
          const value = `${index}`;

          return (
            <Option key={index} value={value}>
              {processDate(startDate, endDate)}
            </Option>
          );
        })}
      </Select>
    );
  };

  const renderCurrentPeriodTextOnly = () => {
    const { startDate, endDate } = periods[0];
    return <strong data-test-id="usage-period-text">{processDate(startDate, endDate)}</strong>;
  };

  return (
    <div className="usage__period-selector">
      <label>API requests usage period</label>
      {isTeamOrEnterpriseCustomer ? renderSelect() : renderCurrentPeriodTextOnly()}
    </div>
  );
};
