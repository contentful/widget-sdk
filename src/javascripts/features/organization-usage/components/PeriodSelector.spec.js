import { shallow } from 'enzyme';
import 'jest-enzyme';
import React from 'react';
import moment from 'moment';
import { Select, Option } from '@contentful/forma-36-react-components';

import { PeriodSelector } from './PeriodSelector';

const DATE_FORMAT = 'YYYY-MM-DD';

const render = (props) => shallow(<PeriodSelector {...props} />);

describe('PeriodSelector', () => {
  let props;

  beforeAll(() => {
    // set fixed date for stable snapshots
    // moment('2017-12-01').unix() = 1512082800
    jest.spyOn(Date, 'now').mockImplementation(() => 1512082800);
  });

  afterAll(() => {
    Date.now.mockRestore();
  });

  beforeEach(() => {
    const startDate = moment().startOf('day').subtract(12, 'days');
    props = {
      periods: [
        {
          sys: { type: 'UsagePeriod', id: 'period1' },
          startDate: startDate.format(DATE_FORMAT),
          endDate: null,
        },
        {
          sys: { type: 'UsagePeriod', id: 'period1' },
          startDate: moment(startDate).subtract(1, 'day').subtract(1, 'month').format(DATE_FORMAT),
          endDate: moment(startDate).subtract(1, 'day').format(DATE_FORMAT),
        },
      ],
      selectedPeriodIndex: 0,
      onChange: jest.fn(),
      isTeamOrEnterpriseCustomer: true,
    };
  });

  it('should render', () => {
    const wrapper = render(props);
    expect(wrapper).toMatchSnapshot();
  });

  it('should call onChange when option is selected', () => {
    const wrapper = render(props);
    expect(props.onChange).not.toHaveBeenCalled();
    wrapper.find(Select).simulate('change', { target: { value: 1 } });
    expect(props.onChange).toHaveBeenCalledWith({ target: { value: 1 } });
  });

  it('should correctly indicate the period is a current period', () => {
    // when endDate is null (default)
    expect(render(props).find(Option).first().html()).toContain('current');

    // when endDate is given (for the community tier)
    props.periods[0].endDate = moment().format(DATE_FORMAT); // today
    expect(render(props).find(Option).first().html()).toContain('current');
  });
});
