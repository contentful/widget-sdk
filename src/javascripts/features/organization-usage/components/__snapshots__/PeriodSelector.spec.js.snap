// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PeriodSelector should render 1`] = `
<div>
  <span
    class="usage__period-selector"
  >
    <label>
      API requests usage period
    </label>
    <div
      class="Select__Select__wrapper___2mbYV Select__Select--auto___3Y-B9"
    >
      <select
        aria-label="period-selector"
        class="Select__Select___2Gi9N a11y__focus-border--default___60AXp"
        data-test-id="period-selector"
        name="period-selector"
      >
        <option
          data-test-id="cf-ui-select-option"
          selected=""
          value="0"
        >
          06 Jan – 05 Feb (current)
        </option>
        <option
          data-test-id="cf-ui-select-option"
          value="1"
        >
          05 Dec – 05 Jan (a month ago)
        </option>
      </select>
      <svg
        class="Icon__Icon___38Epv Icon__Icon--small___1yGZK Icon__Icon--muted___3egnD Select__Select__icon___OBmvS"
        data-test-id="cf-ui-icon"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
    </div>
  </span>
</div>
`;
